### PR TITLE
feat: add `http.redirect_to_https` field to service manifest

### PR DESCRIPTION
**Pull request by dannyrandall**
adds the ability to disable the ALB redirect from http -> https when using a custom domain.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/deploy/svc.go [hunk 1]
```diff
@@ -1360,6 +1360,9 @@ func (d *backendSvcDeployer) validateALBRuntime() error {
 
 func (d *lbWebSvcDeployer) validateALBRuntime() error {
 	hasImportedCerts := len(d.envConfig.HTTPConfig.Public.Certificates) != 0
+	if d.lbMft.RoutingRule.Redirect != nil && *d.lbMft.RoutingRule.Redirect && d.app.Domain == "" && !hasImportedCerts {
```
Comment (paragbhingre) [inline internal/pkg/cli/deploy/svc.go]: What is the difference between 1st two conditions?

`d.lbMft.RoutingRule.Redirect != nil && *d.lbMft.RoutingRule.Redirect`

File: internal/pkg/cli/deploy/svc.go [hunk 2]
```diff
@@ -1360,6 +1360,9 @@ func (d *backendSvcDeployer) validateALBRuntime() error {
...
+		return fmt.Errorf("cannot redirect to https without using a custom domain")
```
Comment (paragbhingre) [inline internal/pkg/cli/deploy/svc.go]: How about?
```suggestion
		return fmt.Errorf(`validate "redirect": cannot redirect to https without using a custom domain`)
```

File: internal/pkg/cli/deploy/svc.go (hunk 1 above)
Comment (dannyrandall) [reply internal/pkg/cli/deploy/svc.go]: `RoutingRule.Redirect` is `*bool`, so it's basically saying "if redirect is configured and set to true". If we didn't have the first condition then the second would panic if `Redirect` was not configured.

File: internal/pkg/cli/deploy/svc_test.go [hunk 3]
```diff
@@ -957,6 +993,58 @@ func TestWorkloadDeployer_DeployWorkload(t *testing.T) {
 				m.mockServiceDeployer.EXPECT().DeployService(gomock.Any(), "mockBucket", gomock.Any()).Return(nil)
 			},
 		},
+		"success with http redirect disabled and certs imported": {
```
Comment (paragbhingre) [inline internal/pkg/cli/deploy/svc_test.go:996]: tests ❤️

File: internal/pkg/cli/deploy/svc.go (hunk 1 above)
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]:
```suggestion
	if aws.BoolValue(d.lbMft.RoutingRule.Redirect) && d.app.Domain == "" && !hasImportedCerts {
```

File: internal/pkg/cli/deploy/svc.go [hunk 4]
```diff
@@ -1360,6 +1360,9 @@ func (d *backendSvcDeployer) validateALBRuntime() error {
 
 func (d *lbWebSvcDeployer) validateALBRuntime() error {
 	hasImportedCerts := len(d.envConfig.HTTPConfig.Public.Certificates) != 0
+	if d.lbMft.RoutingRule.Redirect != nil && *d.lbMft.RoutingRule.Redirect && d.app.Domain == "" && !hasImportedCerts {
+		return fmt.Errorf(`validate "redirect": cannot redirect to https without using a custom domain`)
```
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]:
```suggestion
		return fmt.Errorf(`cannot redirect http to https without having a domain associated with the app %q or importing any certificates in env %q`)
```

File: internal/pkg/deploy/cloudformation/stack/lb_web_svc.go [hunk 5]
```diff
@@ -33,7 +33,8 @@ type loadBalancedWebSvcReadParser interface {
 type LoadBalancedWebService struct {
 	*ecsWkld
 	manifest               *manifest.LoadBalancedWebService
-	httpsEnabled           bool
+	customDomain           bool
+	httpsRedirect          bool
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/lb_web_svc.go]: Why do we need this? Can't we use the manifest field? We only set `HTTPSRedirect` false if it is set `False` in the manifest.

File: internal/pkg/deploy/cloudformation/stack/lb_web_svc.go [hunk 6]
```diff
@@ -66,7 +67,7 @@ type LoadBalancedWebServiceConfig struct {
 func NewLoadBalancedWebService(conf LoadBalancedWebServiceConfig,
 	opts ...LoadBalancedWebServiceOption) (*LoadBalancedWebService, error) {
 	parser := template.New()
-	var dnsDelegationEnabled, httpsEnabled bool
+	var dnsDelegationEnabled, customDomain bool
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/lb_web_svc.go]: This is not customDomain, but strictly means https is enabled...

File: internal/pkg/deploy/cloudformation/stack/lb_web_svc.go [hunk 7]
```diff
@@ -104,6 +105,7 @@ func NewLoadBalancedWebService(conf LoadBalancedWebServiceConfig,
 		},
 		manifest:             conf.Manifest,
 		httpsEnabled:         httpsEnabled,
+		httpRedirect:         httpsEnabled && (conf.Manifest.RoutingRule.Redirect == nil || *conf.Manifest.RoutingRule.Redirect),
```
Comment (iamhopaul123) [inline internal/pkg/deploy/cloudformation/stack/lb_web_svc.go]: maybe consider making `(conf.Manifest.RoutingRule.Redirect == nil || *conf.Manifest.RoutingRule.Redirect)` this a method like `redirectEnabled`

Comment (dannyrandall) [reply internal/pkg/deploy/cloudformation/stack/lb_web_svc.go]: changed the name of the field to `disable_redirect` which i think should help with some confusion!

File: internal/pkg/cli/deploy/svc.go [hunk 8]
```diff
@@ -1360,6 +1360,9 @@ func (d *backendSvcDeployer) validateALBRuntime() error {
 
 func (d *lbWebSvcDeployer) validateALBRuntime() error {
 	hasImportedCerts := len(d.envConfig.HTTPConfig.Public.Certificates) != 0
+	if aws.BoolValue(d.lbMft.RoutingRule.DisableRedirect) && d.app.Domain == "" && !hasImportedCerts {
+		return fmt.Errorf(`validate "redirect": cannot disable http to https redirect without having a domain associated with the app %q or importing any certificates in env %q`, d.app.Name, d.env.Name)
```
Comment (iamhopaul123) [inline internal/pkg/cli/deploy/svc.go]:
```suggestion
		return fmt.Errorf(`cannot disable http to https redirect without having a domain associated with the app %q or importing any certificates in env %q`, d.app.Name, d.env.Name)
```

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/template/templates/workloads/partials/cf/https-listener.yml [hunk 9]
```diff
@@ -85,12 +86,24 @@ HTTPListenerRuleWithDomain:
           Path: "/#{path}"
           Query: "#{query}"
           StatusCode: HTTP_301
+      {{- else}}
+      - TargetGroupArn: !Ref TargetGroup
+        Type: forward
+      {{- end}}
     Conditions:
-{{- if .Aliases }}
+      {{- if .AllowedSourceIps}}
```
Comment (efekarakus) [inline internal/pkg/template/templates/workloads/partials/cf/https-listener.yml:94]: oh is this a bug fix?

File: internal/pkg/deploy/cloudformation/stack/backend_svc.go [hunk 10]
```diff
@@ -152,6 +152,7 @@ func (s *BackendService) Template() (string, error) {
 		Secrets:                  convertSecrets(s.manifest.BackendServiceConfig.Secrets),
 		Aliases:                  aliases,
 		HTTPSListener:            s.httpsEnabled,
+		HTTPRedirect:             s.httpsEnabled,
```
Comment (efekarakus) [inline internal/pkg/deploy/cloudformation/stack/backend_svc.go:155]: gotcha, so we're only supporting this field for now for LBWS as there is no need of a cdn for internal ALBs 👍

**Review by efekarakus — APPROVED** [review]

File: internal/pkg/template/templates/workloads/partials/cf/https-listener.yml (hunk 9 above)
Comment (dannyrandall) [reply internal/pkg/template/templates/workloads/partials/cf/https-listener.yml:94]: "bug" fix, yeah :)

**Pull request merged**